{{ define "notes" }}
<html>
  <head>
    <title>GoDb Search</title>
    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/css/bootstrap.min.css" rel="stylesheet">
    <script src="http://code.jquery.com/jquery-latest.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.16.0/umd/popper.min.js"></script>
    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js"></script>
    <div class="container card text-center shadow p-3 mb-3 bg-light rounded">
        <h2>Godb Genomic Data Notes</h2>
    </div>
    {{ template "navigation" }}
  </head>
  <body>
    <div class="container table-responsive card shadow p-3 mb-1 bg-light rounded">
      <p>Notes on GoDARTS genomic data quality control, assay types and imputation</p>
      General checks on pre-imputation data:
      <ul>
        <li>All data placed on the forward strand and at Genome Reference Consortium Build 37 (GRCh37 / hg19)</li>
        <li>SNPs annotated with dbSNP rsid's (dbSNP 138)</li>
        <li>Variants (markers) checked for call-rate of 95% and above</li>
        <li>Samples checked for call-rates of 95% and above</li>
        <li>Inheritance By Descent checks for relatedness in samples</li>
        <li>Checks for concordance in sex of samples</li>
        <li><a href="/documents/AutomateQC_workflows_update.pdf" download target="_blank">Detailed document (right click)</a></li>
      </ul>
      Final checks prior to imputation:
      <ol>
        <li>Base position and chromosome match with the HRC reference panel</li>
        <li>Strand Check</li>
        <li>Reference allele match</li>
        <li>Remove variants with allele frequency difference > 0.2 between GoDARTS data and reference panel</li>
        <li>Remove palindromic SNPs with Frequency > 0.4</li>
        <li>Remove duplicate markers</li>
        <li>Check indels and remove</li>
      </ol>
      <table id="variantTable" class="table table-striped table-inverse" width="100%" >
        <thead>
          <tr>
            <th>Assay Platform</th>
            <th>Database Tag</th>
            <th>Imputation Panel</th>
            <th>n</th>
            <th>m</th>
            <th>orginal n</th>
            <th>orginal m</th>
            <th>Genotyping Period</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Affymetrix 6 Genome-wide</td><td>affy</td><td>HRC r1.1</td><td>3,884</td><td>39,131,578</td><td>4,134</td><td>932,533</td><td></td>
          </tr>
          <tr>
            <td>Affymetrix 6 Genome-wide</td><td>affy1KG</td><td>1000 Genomes Ph 1</td><td>3,884</td><td>2,748,995*</td><td>4,134</td><td>932,533</td><td></td>
          </tr>
          <tr>
            <td>Illumina Omni Express</td><td>illumina</td><td>HRC r1.1</td><td>3,674</td><td>39,131,578</td><td>3,861</td><td>1,448,315</td><td></td>
          </tr>
          <tr>
            <td>Illumina Omni Express</td><td>illumina1KG</td><td>1000 Genomes Ph 1</td><td>3,674</td><td>2,750,520*</td><td>3,861</td><td>1,448,315</td><td></td>
          </tr>
          <tr>
            <td>Broad Illumina Infinium Custom</td><td>broad</td><td>HRC r1.1</td><td>1,926</td><td>39,131,578</td><td>N/A</td><td>N/A</td><td></td>
          </tr>
          <tr>
            <td>Illumina Exome Array</td><td>exome</td><td>N/A</td><td>4,888</td><td>237,224</td><td>N/A</td><td>247,251</td><td></td>
          </tr>
          <tr>
            <td>Metabochip V2</td><td>metabo</td><td>N/A</td><td>7,179</td><td>178,845</td><td>7,544</td><td>196,725</td><td></td>
          </tr>
          <tr>
            <td>Totals</td><td></td><td></td><td>21,551</td><td>123m+</td><td></td><td></td><td></td>
          </tr></tbody>
          <tr>
            <td>Unique</td><td></td><td></td><td>13,000</td><td>39m+</td><td></td><td></td><td></td>
          </tr></tbody>
      </table>
      * SNPs uniquely present in this imputation panel indexed for use
    </div>
  </body>
</html>
{{ end }}
